File: NR_chap03/extras/backup/poisson_mg_rb.c
#include <stdio.h>             /* A 2D Poisson solver on uniform mesh using Multigrid   */
#include <stdlib.h>            /* and Red/Black RC_Gauss-Seidel (C syntax).                */
#include <math.h>              /* By Paolo Luchini, Thomas Bewley and Anish Karandikar. */
#include <sys/types.h>
#include <time.h>
#include <unistd.h>
#define max(A,B) ( (A) > (B) ? (A) : (B) )
typedef struct {int xbc, ybc, nx, ny, xo, yo, xm, ym; } grid;
void enforce_bcs(double **v, grid g);
void RC_poisson_mg(double **v, double **d, grid g, int n2, int n3);
void RC_poisson_rb(double **v, double **d, grid g);
double max_error(double **v, double **d, grid g);

int main() {
grid g;
int i, j, smooth, s;
double **v, **d, **di, di_sum=0.0, e, o = 0.0;
time_t t0, t1;
/*-------------------------------------- USER INPUT ------------------------------------*/   
/* Take XBC and YBC = 1 for homogeneous Dirichlet (grid:  0:NX   and  0:NY  ),          */
/*                  = 2 for homogeneous Neumann   (grid: -1:NX+1 and -1:NY+1), and      */
/*                  = 3 for periodic              (grid: -1:NX   and -1:NY  ).          */
   const int XBC = 1, YBC = 1, NX = 16, NY = 16, n1=2, n2=2, n3=2;
/*----------------------------------- END OF USER INPUT --------------------------------*/
g.xbc = XBC; g.ybc = YBC; g.nx = NX; g.ny = NY; s = sizeof(double);
printf("Grid: nx = %d, ny = %d \t BCs: (%d, %d)", g.nx, g.ny, g.xbc, g.ybc);
switch(g.xbc) {case 1: g.xm=g.nx+1; g.xo=1; break; 
               case 2: g.xm=g.nx+3; g.xo=2; break;
               case 3: g.xm=g.nx+2; g.xo=2; break; }
switch(g.ybc) {case 1: g.ym=g.ny+1; g.yo=1; break; 
               case 2: g.ym=g.ny+3; g.yo=2; break;
               case 3: g.ym=g.ny+2; g.yo=2; break; }
v =malloc((g.xm+1)*s); for (i=0; i<g.xm; ++i)   v[i] =malloc((g.ym)*s);
d =malloc((g.xm+1)*s); for (i=0; i<g.xm; ++i)   d[i] =malloc((g.ym)*s);
di=malloc((g.xm-3)*s); for (i=0; i<g.xm-4; ++i) di[i]=malloc((g.ym-4)*s);
di_sum = 0.0; for(i=0; i<g.xm-4; ++i) for(j=0; j<g.ym-4;++j) {
      di[i][j] = (((double) rand())/((double) RAND_MAX)) ; di_sum += di[i][j]; }
for(i=0; i<g.xm-4; ++i) free(di[i]); free(di);
for(i=2; i<g.xm-2; ++i) for(j=2; j<=g.ym-3; ++j) 
   d[i][j] = di[i-2][j-2]-di_sum/((double)((g.xm-4)*(g.ym-4)));  
e = max_error(v,d,g);  printf("\nIteration = %d, error = %f\n", 0, e);
for (smooth=1; smooth<=20; ++smooth) RC_poisson_rb(v,d,g);
e=max_error(v,d,g);
printf("After initial smoothing, max defect = %f\n", e);  pause;

t0 = time(NULL); for (i=1; i<=15; ++i) {
   o = e; RC_poisson_mg(v,d,g,n2,n3); e = max_error(v,d,g);
   printf("\nIteration = %d, error = %.10f, factor = %.10f",i,e,o/e);
   if (o/e==1.0) { printf("Converged\n"); break; } } t1 = time(NULL);
printf("Total wall time = %ld sec for %d iterations  -> time/iteration: %f sec\n",
   (t1-t0), i, ((double)(t1-t0)/(double)i));
for(i = 0; i < g.xm; ++i) { free(v[i]); free(d[i]); } free(v); free(d);
} /*------------------------------------------------------------------------------------*/
void enforce_bcs(double **v, grid g) {
int i, m, n; 
switch (g.xbc) {
  case 2: m=g.ym-2; for (i=1;i<=m;++i) {v[0][i]=v[2][i];v[g.xm-1][i]=v[g.xm-3][i];} break;
  case 3: m=g.ym-2; for (i=1;i<=m;++i) {v[0][i]=v[g.xm-2][i];v[g.xm-1][i]=v[1][i];} break;}
switch (g.ybc) {
  case 2: n=g.xm-2; for (i=1;i<=n;++i) {v[i][0]=v[i][2];v[i][g.ym-1]=v[i][g.ym-3];} break;
  case 3: n=g.xm-2; for (i=1;i<=n;++i) {v[i][0]=v[i][g.ym-2];v[i][g.ym-1]=v[i][1];} break;}
} /*------------------------------------------------------------------------------------*/
void RC_poisson_mg(double **vf, double **df, grid gf, int n2, int n3) {
int smooth, i, j, ic, jc, s;
grid gc;
double **vc, **dc;
gc.nx=gf.nx/2; gc.ny=gf.ny/2; gc.xbc=gf.xbc; gc.ybc=gf.ybc;  s=sizeof(double);
switch(gc.xbc) {case 1: gc.xm=gc.nx+1; gc.xo=1; break; 
                case 2: gc.xm=gc.nx+3; gc.xo=2; break;
                case 3: gc.xm=gc.nx+2; gc.xo=2; break; }
switch(gc.ybc) {case 1: gc.ym=gc.ny+1; gc.yo=1; break; 
                case 2: gc.ym=gc.ny+3; gc.yo=2; break;
                case 3: gc.ym=gc.ny+2; gc.yo=2; break; }
for (smooth=1;smooth<=n2;++smooth) RC_poisson_rb(vf,df,gf);
vc=malloc((gc.xm)*s); for (i=0; i<gc.xm; ++i) vc[i]=malloc((gc.ym)*s);
dc=malloc((gc.xm)*s); for (i=0; i<gc.xm; ++i) dc[i]=malloc((gc.ym)*s);
for(i=0; i<gc.xm; ++i) for(j=0; j<gc.ym;++j) { vc[i][j]=0.0; dc[i][j]=0.0; }
for (ic=1;ic<=gc.xm-2;++ic) for (jc=1;jc<=gc.ym-2;++jc) {   /* Compute residual and     */
   i=2*(ic-gc.xo)+gf.xo+1; j=2*(jc-gc.yo)+gf.yo+1;          /* restrict to coarser grid */
   dc[ic][jc]=(vf[i][j+1]+vf[i][j-1]+vf[i+1][j]+vf[i-1][j])*0.25-vf[i][j]+df[i][j];
   vc[ic][jc]=dc[ic][jc];  }
enforce_bcs(vc, gc);                                     
if (gc.nx>3 && gc.ny>3 && (gf.nx % 4)==0 && (gf.ny % 4)==0)    /* Continue to even      */
   RC_poisson_mg(vc,dc,gc,n2,n3);                                 /* coarser grid, or      */
else for (smooth=1;smooth<=20;++smooth) RC_poisson_rb(vc,dc,gc);  /* solve coarsest system */
for (ic=1;ic<=gc.xm-1;++ic) for (jc=1;jc<=gc.ym-1;++jc) {     
   i=2*(ic-gc.xo)+gf.xo+1; j=2*(jc-gc.yo)+gf.yo+1;
   if (j<=gf.ym-1) vf[i-1][j]=vf[i-1][j]+(vc[ic-1][jc]+vc[ic][jc]);     /* Prolongation */
   if (i<=gf.xm-1) vf[i][j-1]=vf[i][j-1]+(vc[ic][jc-1]+vc[ic][jc]); }   /* to fine grid */
enforce_bcs(vf,gf);
for (smooth=1;smooth<=n3;++smooth) RC_poisson_rb(vf,df,gf);
for (i=0; i<gc.xm; ++i) { free(vc[i]); free(dc[i]); }  free(vc); free(dc);
} /*------------------------------------------------------------------------------------*/
void RC_poisson_rb(double **v, double **d, grid g) {
int rb, i, j, m, n;                       /* update red points first, then black points */
for (rb=0; rb<=1; ++rb) { for (i=1; i<=g.xm-2; ++i) { 
   m = 1+(1+i+rb+g.xo+g.yo) % 2; n=g.ym-2;  /* In C, inner loop should be on LAST index */
   for (j=m; j<=n; j+=2) v[i][j]=(v[i][j+1]+v[i][j-1]+v[i+1][j]+v[i-1][j])*0.25+d[i][j]; 
   } enforce_bcs(v,g); }
} /*------------------------------------------------------------------------------------*/
double max_error(double **v, double **d, grid g) {
double e = 0.0;
int i,j;
for (i=1; i<=g.nx-1; ++i) for (j=1; j<=g.ny-1; ++j)  
    e = max(e, fabs((v[i][j+1]+v[i][j-1]+v[i+1][j]+v[i-1][j] )*0.25+d[i][j]-v[i][j] ));
return (e);
}

File: NR_chap03/extras/backup/poisson_mg_rb.m
function RC_poisson_mg_rb
% A 2D Poisson solver on uniform mesh using Multigrid and Red/Black RC_Gauss-Seidel
% (Matlab syntax).  By Paolo Luchini and Thomas Bewley.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% USER INPUT %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Take g.xbc and g.ybc = 1 for homogeneous Dirichlet (grid:  0:nx   and  0:ny  ), 
%                      = 2 for homogeneous Neumann   (grid: -1:nx+1 and -1:ny+1), and
%                      = 3 for periodic              (grid: -1:nx   and -1:ny  ).
plots=1; g.nx=32; g.ny=32;  g.xbc=1;  g.ybc=1;
n1=2; n2=2; n3=2;     % # of rb iterations at various points (take each as 0, 1, or 2)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% END OF USER INPUT %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

fprintf('Grid: nx = %0.6g, ny = %0.6g  BCs: (%0.6g,%0.6g)\n',g.nx,g.ny,g.xbc,g.ybc),
switch g.xbc case 1, g.xo=1; g.xm=g.nx+1;           % Define offset and max grid variables
             case 2, g.xo=2; g.xm=g.nx+3;
             case 3, g.xo=2; g.xm=g.nx+2;  end
switch g.ybc case 1, g.yo=1; g.ym=g.ny+1;
             case 2, g.yo=2; g.ym=g.ny+3;
             case 3, g.yo=2; g.ym=g.ny+2;  end
v=zeros(g.xm,g.ym); d=v;                     % create a random rhs vector d with zero mean
di=rand(g.xm-4,g.ym-4); 
di=di-sum(sum(di))/((g.xm-4)*(g.ym-4)); d(3:g.xm-2,3:g.ym-2)=di;
% for i=3:g.xm-2; for j=3:g.ym-2; d(i,j)=sin((i+j)/10); end; end;
% i=sum(sum(d(:,:)))/((g.xm-4)*(g.ym-4));
% d(3:g.xm-2,3:g.ym-2)=d(3:g.xm-2,3:g.ym-2)-i;
e=max_error(v,d,g,0,2,0,0);
for smooth=1:n1, v=RC_poisson_rb(v,d,g); end                     % n1 iterations of smoothing
tic;
for main=1:150
  o=e; v=RC_poisson_rb(v,d,g); e=max_error(v,d,g,main,2,o,main);

%   o=e; v=RC_poisson_mg(v,d,g,n2,n3); e=max_error(v,d,g,main,plots,o);
%   fprintf('Iteration = %0.6g, error = %0.3e, factor = %0.4f\n',main,e,o/e)
%   if o/e==1, fprintf('Converged\n'), e=max_error(v,d,g,main,1,o);  break, end
end,
t=toc
fprintf('Total time = %0.4g sec for %0.6g iterations',t,main)
fprintf('-> time/iteration: %0.3g sec\n',t/main);        
e=max_error(v,d,g,main,1,o);
% end function RC_poisson_mg_rb
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function v = enforce_bcs(v,g);
% Enforce the Neumann and/or periodic boundary conditions (nothing to do for Dirichlet)
switch  g.xbc  case 2, m=g.ym-1; v(1,2:m)=v(3,     2:m); v(g.xm,2:m)=v(g.xm-2,2:m);
               case 3, m=g.ym-1; v(1,2:m)=v(g.xm-1,2:m); v(g.xm,2:m)=v(2,     2:m); end
switch  g.ybc  case 2, n=g.xm-1; v(2:n,1)=v(2:n,3     ); v(2:n,g.ym)=v(2:n,g.ym-2);
               case 3, n=g.xm-1; v(2:n,1)=v(2:n,g.ym-1); v(2:n,g.ym)=v(2:n,2     ); end     
% end function enforce_bcs
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function vf = RC_poisson_mg(vf,df,gf,n2,n3)
% Apply Multigrid with Red/Black RC_Gauss-Seidel smoothing to the Poisson equation.
gc.nx=gf.nx/2;  gc.ny=gf.ny/2;  gc.xbc=gf.xbc;  gc.ybc=gf.ybc;        % Define coarse grid
switch gc.xbc case 1, gc.xm=gc.nx+1; gc.xo=1;
              case 2, gc.xm=gc.nx+3; gc.xo=2;
              case 3, gc.xm=gc.nx+2; gc.xo=2;  end
switch gc.ybc case 1, gc.ym=gc.ny+1; gc.yo=1;
              case 2, gc.ym=gc.ny+3; gc.yo=2;
              case 3, gc.ym=gc.ny+2; gc.yo=2;  end
% gf.xm, gf.ym, df, vf
for smooth=1:n2, vf=RC_poisson_rb(vf,df,gf); end          % n2 iterations of rb smoothing
vc=zeros(gc.xm,gc.ym);  dc=zeros(gc.xm,gc.ym);         % Initialize coarser grid variables
for ic=2:gc.xm-1                          % Calculate residual and perform the restriction      
   for jc=2:gc.ym-1                       % (using direct injection) in a single step.
      i=2*(ic-gc.xo)+gf.xo; j=2*(jc-gc.yo)+gf.yo; 
      dc(ic,jc)=((vf(i,j+1)+vf(i,j-1)+vf(i+1,j)+vf(i-1,j))*0.25-vf(i,j)+df(i,j));
      vc(ic,jc)=dc(ic,jc);       % Paolo's trick: this is a better guess than 0 to init vc
   end
end
% gc.xm, dc, vc, pause;
vc=enforce_bcs(vc,gc);
if gc.nx > 3 & gc.ny > 3 & mod(gf.nx,4)==0 & mod(gf.ny,4)==0 
   vc=RC_poisson_mg(vc,dc,gc,n2,n3);                 % Continue down to an even coarser grid,
else                                              % or
   for loop=1:20, vc=RC_poisson_rb(vc,dc,gc); end    % solve coarsest system (almost exactly)
end
for ic=2:gc.xm       % Prolongation (bilinear interpolation) on the black interior points.
   for jc=2:gc.ym    % Note: the next call to RC_poisson_rb will take care of the red points.
      i=2*(ic-gc.xo)+gf.xo; j=2*(jc-gc.yo)+gf.yo;
      if j<=gf.ym, vf(i-1,j)=vf(i-1,j)+(vc(ic-1,jc)+vc(ic,jc)); end
      if i<=gf.xm, vf(i,j-1)=vf(i,j-1)+(vc(ic,jc-1)+vc(ic,jc)); end
   end
end
vf=enforce_bcs(vf,gf);
for smooth=1:n3, vf=RC_poisson_rb(vf,df,gf); end                 % n3 iterations of smoothing
% end function RC_poisson_mg
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function v = RC_poisson_rb(v,d,g);
% Apply Red/Black RC_Gauss-Seidel smoothing, with L derived from the Poisson equation
% (normalized s.t. the diagonal elements of L are 1) on the grid g, to Lv=d.
for rb=0:1,  for i=2:g.xm-1                  % update red points first, then black points.
   m=2+mod(i+rb+g.xo+g.yo,2); n=g.ym-1;
   mp=m+1; mm=m-1; np=n+1; nm=n-1; ip=i+1; im=i-1;
   v(i,m:2:n)=(v(i,mp:2:np)+v(i,mm:2:nm)+v(ip,m:2:n)+v(im,m:2:n))*0.25+d(i,m:2:n);
end, v=enforce_bcs(v,g); end
% end function RC_poisson_rb
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function e = max_error(v,d,g,main,plots,iter)
% Compute the maximum error and, if requested, plot both the solution and the error.
e=0;
for i=2:g.nx, for j=2:g.ny
   e=max(e,abs((v(i,j+1)+v(i,j-1)+v(i+1,j)+v(i-1,j))*0.25+d(i,j)-v(i,j)));
end, end
if plots>0,
   figure(1);  clf; surf(v(g.xo:g.xo+g.nx,g.yo:g.yo+g.ny)); 
   title(sprintf('Solution at iteration = %0.6g',main))
   for i=2:g.xm-1, for j=2:g.ym-1
      error(i,j)=((v(i,j+1)+v(i,j-1)+v(i+1,j)+v(i-1,j))*0.25+d(i,j)-v(i,j));
   end, end
   figure(2);  clf; surf(error(2:2:g.xm-1,2:2:g.ym-1));
   title(sprintf('Error at iteration = %0.6g',main))
   if plots>1, fn=['err' num2str(iter) '.eps'];  print('-depsc',fn); end
end
% end function max_error
